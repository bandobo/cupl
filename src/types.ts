export interface Node {
  type: 'condition' | 'activity' | 'action' | 'start' | 'stop' | 'end';
  name?: string;
}

export interface IfNode {
  type: 'if';
  branches: Branch[];
}

export interface Diagram {
  title: string;
  description?: string;
  nodes: (Node | IfNode)[];
}

export interface Config {
  paths?: { [key: string]: PathConfig };
  global?: GlobalConfig;
}

export interface Branch {
  node: Node;
  children: Node[];
}

export interface CommonConfig {
  alias?: { [key: string]: string };
  examples?: string[][];
}

export interface PathConfig extends CommonConfig {
  name?: string;
  tags?: string[];
}
export interface GlobalConfig extends CommonConfig {
  indent?: number;
  dialect?: string;
}
