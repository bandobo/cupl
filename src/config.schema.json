{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$ref": "#/definitions/Config",
  "definitions": {
    "Config": {
      "type": "object",
      "properties": {
        "paths": {
          "type": "object",
          "additionalProperties": {
            "$ref": "#/definitions/PathConfig"
          }
        },
        "global": {
          "$ref": "#/definitions/GlobalConfig"
        }
      },
      "additionalProperties": false
    },
    "PathConfig": {
      "type": "object",
      "properties": {
        "alias": {
          "type": "object",
          "additionalProperties": {
            "type": "string"
          }
        },
        "examples": {
          "anyOf": [
            {
              "type": "array",
              "items": {
                "type": "object",
                "additionalProperties": {
                  "type": "string"
                }
              }
            },
            {
              "type": "array",
              "items": {
                "type": "array",
                "items": {
                  "type": "string"
                }
              }
            }
          ]
        },
        "name": {
          "type": "string"
        },
        "tags": {
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      },
      "additionalProperties": false
    },
    "GlobalConfig": {
      "type": "object",
      "properties": {
        "alias": {
          "type": "object",
          "additionalProperties": {
            "type": "string"
          }
        },
        "examples": {
          "anyOf": [
            {
              "type": "array",
              "items": {
                "type": "object",
                "additionalProperties": {
                  "type": "string"
                }
              }
            },
            {
              "type": "array",
              "items": {
                "type": "array",
                "items": {
                  "type": "string"
                }
              }
            }
          ]
        },
        "indent": {
          "type": "number"
        },
        "dialect": {
          "type": "string"
        }
      },
      "additionalProperties": false
    }
  }
}
