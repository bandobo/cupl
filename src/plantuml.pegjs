PlantUMLFile
  = diagrams:Diagrams
  {
    return diagrams;
  }
  / (!"@startuml" .)*
  {
    return []
  }

Diagrams
  = diagrams:(
    (!"@startuml" .)*
    _ "@startuml" _ [^\n]* (_ NewLine) +
    title:(Title)?
    nodes:Sequence
    _ "@enduml" _ NewLine?
    (!"@startuml" .)*
    {
      return {
        ...title,
        nodes,
      };
    }
  )+

Title =
  title:(ShortTitle / LongTitle)
  {
    return title;
  }

ShortTitle =
  _ "title" _ title:[^\n]+ (_ NewLine) +
  {
    return {
      title: title.join('').trim(),
    }
  }

LongTitle =
  _ "title" _ NewLine _ longTitle:( !(_ "end" _ "title") [^\n]* NewLine)+ _ "end" _ "title" (_ NewLine)+
  {
    const [title, ...description] = longTitle;
    return {
      title: title[1].join('').trim(),
      description: description.map((l:string[][]) => l[1].join('').trim()).join('\n'),
    }
  }


Sequence =
  sequence:(Node / If)*
  {
    return sequence;
  }

If =
  _ "if" firstBranch:IfTitle firstChildren:Sequence elseIf:ElseIf* _ "else" lastBranch:IfTail lastChildren:Sequence EndIf
  {
    return {
      type: 'if',
      branches: [
        {node: firstBranch, children: firstChildren},
        ...elseIf,
        {node: lastBranch, children: lastChildren},
      ],
    };
  }

Node = _ node:(Terminal / Activity) _ NewLine
  {
    return node;
  }

ElseIf =
  _ "elseif" _ node:IfTitle children:Sequence
  {
    return { node, children };
  }
EndIf =
  _ "endif" _ NewLine


IfTitle =
  _ "(" name:IfText ")" _  "then" IfTail
  {
    return {
      type: 'condition',
      name,
    };
  }

IfTail = _ "("  name:IfText ")" _ NewLine
  {
    return { name, type: 'condition' }
  }

IfText =
  text:( !([?][)]) [^)] )+ "?"?
  {
    return text.map((t: string[]) => t[1]).join('')
  }

Terminal =
  type:(Start / Stop / End)
  {
    return { type } as types.Node;
  }

Start = "start"
Stop = "stop"
End = "end"

Activity =
  ActivityStart activity:(([@])?[^;|<>/\]}]+) ActivityEnd Note?
  {
    return {
      name: activity[1].join('').replace(/\n\s+/g, '\n'),
      type: activity[0] ? 'action' : 'activity'
    } as types.Node;
  }

Note =
  _ NewLine _ "note" _ ("left"/"right")_ NewLine _ ( !(_ "end" _ "note") [^\n]* NewLine)+ _ "end" _ "note" _

ActivityStart = ":"
ActivityEnd =
  ";"
  / "|"
  / "<"
  / ">"
  / "/"
  / "]"
  / "}"

// =======
Name
  = name:([A-Za-z0-9._]+)
  {
    return name.join('');
  }

DiagramId
  = "(" _ "id" _ "=" Name ")"
_
  = [ \t]*

NewLine
  = "\n"
  / "\r\n"
