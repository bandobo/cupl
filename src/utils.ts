import { Predicate } from 'fp-ts/lib/function';

export const split = <T>(predicate: Predicate<T>) => (as: T[]) => {
  const result: T[][] = [];
  let current: T[] = [];
  as.forEach((e) => {
    if (!predicate(e)) {
      current.push(e);
    } else if (current.length > 0) {
      result.push(current);
      current = [];
    }
  });
  if (current.length > 0) {
    result.push(current);
  }
  return result;
};

export const spaces = ((n: number) => {
  const res: string[] = [];
  for (let index = 0; index < n; index++) {
    res.push(index === 0 ? '' : res[index - 1] + ' ');
  }
  return res;
})(20);

export const deepFlatten = (a: any[]) => {
  let res: any[] = [];
  a.forEach((item) => {
    if (item.length === undefined) {
      res.push(item);
    } else {
      res = res.concat(deepFlatten(item));
    }
  });
  return res;
};
