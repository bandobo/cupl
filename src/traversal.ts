import { IfNode, Node } from './types';
import { deepFlatten } from './utils';

export function* traverse(diagram: (Node | IfNode)[]) {
  const traversal: any = {};
  function* visitBranches(node: IfNode) {
    for (const branch of node.branches) {
      for (const [path, terminated] of traversal.visitSequence(branch.children)) {
        yield [[branch.node, ...path], terminated];
      }
    }
  }
  function* visitSequence(nodes: (Node | IfNode)[]): Generator<[Node[], boolean], void, unknown> {
    const segments: any[] = [];
    let paths: Node[][] = [];
    const path: Node[] = [];
    function visitSegment(idx: number) {
      if (idx >= segments.length) {
        paths.push([...path]);
        return;
      }
      for (const subPath of segments[idx]) {
        path.push(subPath);
        visitSegment(idx + 1);
        path.pop();
      }
    }

    for (const node of nodes) {
      const terminatedBranches = [];
      const onGoingBranches = [];
      if (node.type === 'if') {
        for (const [branch, isTerminated] of traversal.visitBranches(node)) {
          if (isTerminated) {
            terminatedBranches.push(branch);
          } else {
            onGoingBranches.push(branch);
          }
        }
      } else if (['end', 'stop'].includes(node.type)) {
        terminatedBranches.push(node);
      } else {
        onGoingBranches.push(node);
      }
      if (terminatedBranches.length > 0) {
        segments.push(terminatedBranches);
        paths = [];
        visitSegment(0);
        for (const path of paths) {
          yield [path, true];
        }
        segments.pop();
      }
      if (onGoingBranches.length > 0) {
        segments.push(onGoingBranches);
      } else {
        return;
      }
    }
    paths = [];
    visitSegment(0);
    for (const path of paths) {
      yield [path, false];
    }
  }
  traversal.visitBranches = visitBranches;
  traversal.visitSequence = visitSequence;
  for (const [path] of visitSequence(diagram)) {
    yield deepFlatten(path);
  }
}
