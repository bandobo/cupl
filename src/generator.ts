import { dropLeft, filterMap, findLastIndex, last } from 'fp-ts/lib/Array';
import { flow } from 'fp-ts/lib/function';
import { getOrElse, none, some, toNullable } from 'fp-ts/lib/Option';
import gherkin from 'gherkin';
import * as _ from 'lodash/fp';

import { traverse } from './traversal';
import { Config, Diagram, GlobalConfig, Node, PathConfig } from './types';
import { spaces, split } from './utils';

const getOrZero = getOrElse<number>(() => 0);

const findLastConditionIndex = findLastIndex<Node>((n) => n.type === 'condition');

// const inspect = <T>() => (a: T) => {
//   // eslint-disable-next-line no-console
//   console.log(a);
//   return a;
// };

const DEFAULT_PATH_CONFIG: PathConfig = {
  tags: [],
};
const DEFAULT_DIALECT = 'en';

const is = (type: Node['type']) => (n: Node) => n.type === type;
// const optionalIs = (type: Node['type']) => (n: Node) => (n.type === type ? some(n) : none);
const returnSomeNameIfTypeIs = (type: Node['type']) => (n: Node) =>
  n.type === type ? some(n.name!) : none;

const nullableLastString: <T extends string>(as: [T]) => T | null = flow(last, toNullable);
const lastSegmentOrEmpty = <T extends string>(as: T[][]) => as[as.length - 1] || [];

const removeNewLines = (name: string) => name.replace(/\s*(\n|\r\n)\s*/g, ' ');

const getAlias = (pathConfig?: PathConfig, globalConfig?: GlobalConfig) => (givenName: string) => {
  const name = removeNewLines(givenName);
  const globalAlias = globalConfig?.alias?.[name];
  const pathAlias = pathConfig?.alias?.[name];
  return removeNewLines(pathAlias || globalAlias || name);
};

const INDENT_SIZE = 2;

const getDialect = (globalConfig?: GlobalConfig) => {
  const dialects = gherkin.dialects();
  return dialects[globalConfig?.dialect || DEFAULT_DIALECT] || dialects[DEFAULT_DIALECT];
};

const getDialectCode = (globalConfig?: GlobalConfig) => {
  const dialects = gherkin.dialects();
  return dialects[globalConfig?.dialect || 'unknown'] ? globalConfig?.dialect : DEFAULT_DIALECT;
};

const getStep = (indentSize: number) => (type: string, content: string, indent?: boolean) =>
  `${spaces[indentSize * (indent ? 3 : 2)]}${type}${content}`;

const getPathKey = (path: Node[]) =>
  path
    .filter((n) => n.type === 'condition')
    .map((t) => removeNewLines(t.name!))
    .join('|');

export function generateScenario(
  path: Node[],
  config?: { path?: PathConfig; global?: GlobalConfig }
) {
  const indentSize = config?.global?.indent || INDENT_SIZE;
  const alias = getAlias(config?.path, config?.global);
  const step = getStep(indentSize);
  const scenario = config?.path?.name || getPathKey(path);
  const dialect = getDialect(config?.global);
  const tags = config?.path?.tags || [];
  const givenConditions = flow(filterMap(returnSomeNameIfTypeIs('condition')))(path);
  const whenActions = flow(
    split(is('condition')),
    filterMap((segment: Node[]) => {
      const remain = filterMap(returnSomeNameIfTypeIs('action'))(segment);
      return remain.length > 0 ? some(remain) : none;
    }),
    lastSegmentOrEmpty
    // getOrElse(() => [])
  )(path);
  const lastConditionIndex = flow(findLastConditionIndex, getOrZero)(path);
  const thenActivities = filterMap(returnSomeNameIfTypeIs('activity'))(
    dropLeft(lastConditionIndex)(path)
  );

  const vocab = {
    scenario: nullableLastString(dialect.scenario as any)!,
    and: nullableLastString(dialect.and as any)!,
    then: nullableLastString(dialect.then as any)!,
    when: nullableLastString(dialect.when as any)!,
    given: nullableLastString(dialect.given as any)!,
  };

  return [
    ...(tags.length > 0 ? [`${spaces[indentSize]}${tags.map((tag) => `@${tag}`).join(' ')}`] : []),
    `${spaces[indentSize]}${vocab.scenario}: ${scenario}`,
    ...givenConditions.map((name: string, idx: number) =>
      step(idx === 0 ? vocab.given : vocab.and, alias(name), idx !== 0)
    ),
    ...whenActions.map((name: string, idx: number) =>
      step(idx === 0 ? vocab.when : vocab.and, alias(name), idx !== 0)
    ),
    ...thenActivities.map((name: string, idx: number) =>
      step(idx === 0 ? vocab.then : vocab.and, alias(name), idx !== 0)
    ),
  ];
}

export function generateFeature(diagram: Diagram, config: Config) {
  const paths = [...traverse(diagram.nodes)].map((path) => ({
    nodes: path,
    key: getPathKey(path),
  }));
  const dialect = getDialect(config?.global);
  const dialectCode = getDialectCode(config?.global);
  return {
    feature:
      (dialectCode === DEFAULT_DIALECT ? '' : `# language: ${dialectCode}\n`) +
      `${dialect.feature[0]}: ${diagram.title}` +
      (diagram.description ? ('\n' + diagram.description).replace(/\n/g, '\n  ') + '\n' : '') +
      '\n' +
      paths
        .map((path) =>
          generateScenario(path.nodes, {
            path: config.paths?.[path.key],
            global: config.global,
          }).join('\n')
        )
        .join('\n\n'),
    newConfig: {
      $schema: 'https://raw.githubusercontent.com/cinoss/cupl/master/src/config.schema.json',
      ...config,
      global: {
        dialect: dialectCode,
        ...config.global,
      },
      paths: {
        ..._.flow(
          _.map((path: { key: string }) => [path.key, DEFAULT_PATH_CONFIG]),
          _.fromPairs
        )(paths),
        ...config.paths,
      },
    },
  };
}
