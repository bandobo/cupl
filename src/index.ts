import { Command, flags } from '@oclif/command';
import { green } from 'chalk';
import { existsSync, readFileSync, writeFileSync } from 'fs';
import * as stringify from 'json-stable-stringify';
import { join, parse as parsePath } from 'path';

import { generateFeature } from './generator';
import { parse } from './plantuml';

const CONFIG_EXT = 'cupl.json';
const FEATURE_EXT = 'feature';

class Cupl extends Command {
  static description =
    'CLI tool for automatic CUcumber gherkin feature files generation from PLantuml activity diagram.';

  static flags = {
    // add --version flag to show CLI version
    version: flags.version({ char: 'v' }),
    help: flags.help({ char: 'h' }),
    write: flags.boolean({
      char: 'w',
      description: 'write generated output to a file with .feature extension',
    }),
  };

  static args = [{ name: 'file' }];

  async run() {
    const { args, flags } = this.parse(Cupl);
    const puml = parse(readFileSync(args.file, 'utf-8'), { filename: args.file });
    const path = parsePath(args.file);
    const configPath = join(path.dir, `${path.name}.${CONFIG_EXT}`);
    const featurePath = join(path.dir, `${path.name}.${FEATURE_EXT}`);
    const config = existsSync(configPath) ? JSON.parse(readFileSync(configPath, 'utf-8')) : {};
    const { feature, newConfig } = generateFeature(puml[0], config);
    this.log(feature);
    writeFileSync(configPath, stringify(newConfig, { space: '  ' }));
    if (flags.write) {
      writeFileSync(featurePath, feature);
      this.log(`\nGenerated ${green(featurePath)} successfully!`);
    }
  }
}

export = Cupl;
