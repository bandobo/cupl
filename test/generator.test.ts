import { generateFeature, generateScenario } from '../src/generator';
import { parse } from './../src/plantuml';

describe('generateScenario', () => {
  // async function assertIsValidGherkin(scenario: string[]) {
  //   // const output = gherkin.fromSources([makeSourceEnvelop(scenario.join('\n'))]);
  //   console.log(gherkin.dialects().en);
  // }
  test('should create scenario with Given Then When', async () => {
    const generated = generateScenario([
      { type: 'start' },
      { type: 'action', name: 'user enters PIN' },
      { type: 'condition', name: 'entered wrong PIN' },
      { type: 'activity', name: 'show wrong PIN Entered' },
      { type: 'activity', name: 'eject card' },
      { type: 'stop' },
    ]);
    expect(generated).toMatchSnapshot();
    expect(generated.find((line) => /Scenario/.test(line))).toBeTruthy();
    expect(generated.find((line) => /Given/.test(line))).toBeTruthy();
    expect(generated.find((line) => /When/.test(line))).toBeTruthy();
    expect(generated.find((line) => /Then/.test(line))).toBeTruthy();
    // await assertIsValidGherkin(generated);
  });
  test('should accept path with no condition', () => {
    const generated = generateScenario([
      { type: 'start' },
      { type: 'action', name: 'user enters PIN' },
      { type: 'activity', name: 'show wrong PIN Entered' },
      { type: 'activity', name: 'eject card' },
      { type: 'stop' },
    ]);
    expect(generated).toMatchSnapshot();
    expect(generated.find((line) => /Scenario/.test(line))).toBeTruthy();
    expect(generated.find((line) => /Given/.test(line))).toBeFalsy();
    expect(generated.find((line) => /When/.test(line))).toBeTruthy();
    expect(generated.find((line) => /Then/.test(line))).toBeTruthy();
  });
  test('should accept path with neither action nor condition', () => {
    const generated = generateScenario([
      { type: 'start' },
      { type: 'activity', name: 'show wrong PIN Entered' },
      { type: 'activity', name: 'eject card' },
      { type: 'stop' },
    ]);
    expect(generated).toMatchSnapshot();
    expect(generated.find((line) => /Scenario/.test(line))).toBeTruthy();
    expect(generated.find((line) => /Given/.test(line))).toBeFalsy();
    expect(generated.find((line) => /When/.test(line))).toBeFalsy();
    expect(generated.find((line) => /Then/.test(line))).toBeTruthy();
  });
  test('should support And on Given, When and Then', () => {
    const generated = generateScenario([
      { type: 'start' },
      { type: 'action', name: 'user push a card in' },
      { type: 'action', name: 'user enters PIN' },
      { type: 'condition', name: 'server responded' },
      { type: 'condition', name: 'entered wrong PIN' },
      { type: 'activity', name: 'show wrong PIN Entered' },
      { type: 'activity', name: 'eject card' },
      { type: 'stop' },
    ]);
    expect(generated).toMatchSnapshot();
    expect(generated.filter((line) => /^\s+And/.test(line)).length).toEqual(3);
  });
  describe('config', () => {
    test('support path and global alias', () => {
      const generated = generateScenario(
        [
          { type: 'start' },
          { type: 'action', name: 'user enters PIN' },
          { type: 'condition', name: 'entered wrong PIN' },
          { type: 'activity', name: 'show wrong PIN Entered' },
          { type: 'activity', name: 'eject card' },
          { type: 'stop' },
        ],
        {
          path: {
            name: 'User forgot PIN or made a typo',
            tags: ['web', 'server'],
            alias: {
              'entered wrong PIN': 'having 123456 as PIN',
              'user enters PIN': 'enter 111111 as PIN',
            },
          },
          global: {
            alias: {
              'entered wrong PIN': '?????????',
              'user enters PIN': '????????',
              'eject card': 'return card to user',
            },
          },
        }
      );
      expect(generated).toMatchSnapshot();
      expect(generated.find((line) => /User forgot PIN or made a typo/.test(line))).toBeTruthy();
      expect(generated.find((line) => /123456/.test(line))).toBeTruthy();
      expect(generated.filter((line) => /\?+/.test(line)).length).toEqual(0);
    });
    test('support dialects', () => {
      const generated = generateScenario(
        [
          { type: 'start' },
          { type: 'action', name: 'user enters PIN' },
          { type: 'condition', name: 'entered wrong PIN' },
          { type: 'activity', name: 'show wrong PIN Entered' },
          { type: 'activity', name: 'eject card' },
          { type: 'stop' },
        ],
        {
          path: {
            name: 'User forgot PIN or made a typo',
            tags: ['web', 'server'],
            alias: {
              'entered wrong PIN': 'having 123456 as PIN',
              'user enters PIN': 'enter 111111 as PIN',
            },
          },
          global: {
            dialect: 'vi',
            alias: {
              'entered wrong PIN': '?????????',
              'user enters PIN': '????????',
              'eject card': 'return card to user',
            },
          },
        }
      );
      expect(generated).toMatchSnapshot();
      expect(generated.find((line) => /User forgot PIN or made a typo/.test(line))).toBeTruthy();
      expect(generated.find((line) => /123456/.test(line))).toBeTruthy();
      expect(generated.filter((line) => /\?+/.test(line)).length).toEqual(0);
    });
    test('support fallback to en if specified dialect is missing', () => {
      const generated = generateScenario(
        [
          { type: 'start' },
          { type: 'action', name: 'user enters PIN' },
          { type: 'condition', name: 'entered wrong PIN' },
          { type: 'activity', name: 'show wrong PIN Entered' },
          { type: 'activity', name: 'eject card' },
          { type: 'stop' },
        ],
        {
          path: {
            name: 'User forgot PIN or made a typo',
            tags: ['web', 'server'],
            alias: {
              'entered wrong PIN': 'having 123456 as PIN',
              'user enters PIN': 'enter 111111 as PIN',
            },
          },
          global: {
            dialect: 'alien',
            alias: {
              'entered wrong PIN': '?????????',
              'user enters PIN': '????????',
              'eject card': 'return card to user',
            },
          },
        }
      );
      expect(generated).toMatchSnapshot();
      expect(generated.find((line) => /\s+Given/.test(line))).toBeTruthy();
    });
    it('should support scenario examples', () => {
      const generated = generateScenario(
        [
          { type: 'start' },
          { type: 'action', name: 'user enters PIN' },
          { type: 'condition', name: 'entered wrong PIN' },
          { type: 'activity', name: 'show wrong PIN Entered' },
          { type: 'activity', name: 'eject card' },
          { type: 'stop' },
        ],
        {
          path: {
            name: 'User forgot PIN or made a typo',
            tags: ['web', 'server'],
            alias: {
              'entered wrong PIN': 'having <secret> as PIN',
              'user enters PIN': 'enter <input> as PIN',
            },
            examples: [
              ['secret', 'input'],
              ['111111', ''],
              ['111111', '111112'],
            ],
          },
          global: {
            dialect: 'alien',
            alias: {
              'entered wrong PIN': '?????????',
              'user enters PIN': '????????',
              'eject card': 'return card to user',
            },
          },
        }
      );
      expect(generated).toMatchSnapshot();
      expect(generated.filter((line) => /secret/.test(line))).toHaveLength(2);
    });
    it('should throw an error if header row has no items', () => {
      const exec = () => {
        generateScenario(
          [
            { type: 'start' },
            { type: 'action', name: 'user enters PIN' },
            { type: 'condition', name: 'entered wrong PIN' },
            { type: 'activity', name: 'show wrong PIN Entered' },
            { type: 'activity', name: 'eject card' },
            { type: 'stop' },
          ],
          {
            path: {
              name: 'User forgot PIN or made a typo',
              tags: ['web', 'server'],
              alias: {
                'entered wrong PIN': 'having <secret> as PIN',
                'user enters PIN': 'enter <input> as PIN',
              },
              examples: [[], ['111111', ''], ['111111', '111112']],
            },
            global: {
              dialect: 'alien',
              alias: {
                'entered wrong PIN': '?????????',
                'user enters PIN': '????????',
                'eject card': 'return card to user',
              },
            },
          }
        );
      };
      expect(exec).toThrowError();
    });
    it('should throw an error if a header is empty', () => {
      const exec = () => {
        generateScenario(
          [
            { type: 'start' },
            { type: 'action', name: 'user enters PIN' },
            { type: 'condition', name: 'entered wrong PIN' },
            { type: 'activity', name: 'show wrong PIN Entered' },
            { type: 'activity', name: 'eject card' },
            { type: 'stop' },
          ],
          {
            path: {
              name: 'User forgot PIN or made a typo',
              tags: ['web', 'server'],
              alias: {
                'entered wrong PIN': 'having <secret> as PIN',
                'user enters PIN': 'enter <input> as PIN',
              },
              examples: [[''], ['111111', ''], ['111111', '111112']],
            },
            global: {
              dialect: 'alien',
              alias: {
                'entered wrong PIN': '?????????',
                'user enters PIN': '????????',
                'eject card': 'return card to user',
              },
            },
          }
        );
      };
      expect(exec).toThrowError();
    });
  });
});

describe('generateFeature', () => {
  test('Should support no description', () => {
    const diagrams = parse(`
      @startuml
      title title
      start
      :Hello world;
      :This is defined on

      several **lines**;
      end
      @enduml
    `);
    expect(generateFeature(diagrams[0], {})).toMatchSnapshot();
  });
  test('Should support long description', () => {
    const diagrams = parse(`
      @startuml
      title
      title
      description
      description
      description
      description
      end title
      start
      :Hello world;
      :This is defined on

      several **lines**;
      end
      @enduml
    `);
    expect(generateFeature(diagrams[0], {})).toMatchSnapshot();
  });
  test('Should include language tag on non default dialect', () => {
    const diagrams = parse(`
      @startuml
      title
      title
      description
      description
      description
      description
      end title
      start
      :Hello world;
      :This is defined on

      several **lines**;
      end
      @enduml
    `);
    expect(generateFeature(diagrams[0], { global: { dialect: 'vi' } })).toMatchSnapshot();
  });
  test('Should support assert (!)', () => {
    const diagrams = parse(`
      @startuml
      start
      :Hello world;
      :this is condition!;
      end
      @enduml
    `);
    const { feature } = generateFeature(diagrams[0], {});
    expect(feature).toMatchSnapshot();
    expect(feature).toContain('Given this is condition');
  });
  test('Should support empty if branch', () => {
    const diagrams = parse(`
      @startuml
      start
      :Hello world;

      if (a) then (b)
      else (c)
      endif

      end
      @enduml
    `);
    const { feature } = generateFeature(diagrams[0], {});
    expect(feature).toMatchSnapshot();
  });
});
