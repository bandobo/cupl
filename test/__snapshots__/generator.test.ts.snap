// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateFeature Should include language tag on non default dialect 1`] = `
Object {
  "feature": "# language: vi
Tính năng: title
  description
  description
  description
  description

  Kịch bản: 
    Thì Hello world
      Và This is defined on several **lines**",
  "newConfig": Object {
    "$schema": "https://raw.githubusercontent.com/cinoss/cupl/master/src/config.schema.json",
    "global": Object {
      "dialect": "vi",
    },
    "paths": Object {
      "": Object {},
    },
  },
}
`;

exports[`generateFeature Should support long description 1`] = `
Object {
  "feature": "Feature: title
  description
  description
  description
  description

  Scenario: 
    Then Hello world
      And This is defined on several **lines**",
  "newConfig": Object {
    "$schema": "https://raw.githubusercontent.com/cinoss/cupl/master/src/config.schema.json",
    "global": Object {
      "dialect": "en",
    },
    "paths": Object {
      "": Object {},
    },
  },
}
`;

exports[`generateFeature Should support no description 1`] = `
Object {
  "feature": "Feature: title
  Scenario: 
    Then Hello world
      And This is defined on several **lines**",
  "newConfig": Object {
    "$schema": "https://raw.githubusercontent.com/cinoss/cupl/master/src/config.schema.json",
    "global": Object {
      "dialect": "en",
    },
    "paths": Object {
      "": Object {},
    },
  },
}
`;

exports[`generateScenario should accept path with neither action nor condition 1`] = `
Array [
  "  Scenario: ",
  "    Then show wrong PIN Entered",
  "      And eject card",
]
`;

exports[`generateScenario should accept path with no condition 1`] = `
Array [
  "  Scenario: ",
  "    When user enters PIN",
  "    Then show wrong PIN Entered",
  "      And eject card",
]
`;

exports[`generateScenario should create scenario with Given Then When 1`] = `
Array [
  "  Scenario: entered wrong PIN",
  "    Given entered wrong PIN",
  "    When user enters PIN",
  "    Then show wrong PIN Entered",
  "      And eject card",
]
`;

exports[`generateScenario should support And on Given, When and Then 1`] = `
Array [
  "  Scenario: server responded|entered wrong PIN",
  "    Given server responded",
  "      And entered wrong PIN",
  "    When user push a card in",
  "      And user enters PIN",
  "    Then show wrong PIN Entered",
  "      And eject card",
]
`;

exports[`generateScenario support dialects 1`] = `
Array [
  "  @web @server",
  "  Kịch bản: User forgot PIN or made a typo",
  "    Cho having 123456 as PIN",
  "    Khi enter 111111 as PIN",
  "    Thì show wrong PIN Entered",
  "      Và return card to user",
]
`;

exports[`generateScenario support fallback to en if specified dialect is missing 1`] = `
Array [
  "  @web @server",
  "  Scenario: User forgot PIN or made a typo",
  "    Given having 123456 as PIN",
  "    When enter 111111 as PIN",
  "    Then show wrong PIN Entered",
  "      And return card to user",
]
`;

exports[`generateScenario support path and global alias 1`] = `
Array [
  "  @web @server",
  "  Scenario: User forgot PIN or made a typo",
  "    Given having 123456 as PIN",
  "    When enter 111111 as PIN",
  "    Then show wrong PIN Entered",
  "      And return card to user",
]
`;
